src/perf/lib/ThroughputClient.cpp

```diff
@@ -0,0 +1,419 @@
+        MsQuic->ConnectionOpen(
+            Session,
+            [](HQUIC Handle, void* Context, QUIC_CONNECTION_EVENT* Event) -> QUIC_STATUS {
+                ConnectionData* ConnData = (ConnectionData*)Context;
+                return ConnData->Client->
+                    ConnectionCallback(
+                        Handle,
+                        Event,
+                        ConnData);
+            },
+            ConnData,
+            &ConnData->Connection.Handle);
+
+    if (QUIC_FAILED(Status)) {
+        WriteOutput("Failed ConnectionOpen 0x%x\n", Status);
+        ConnectionDataAllocator.Free(ConnData);
+        return Status;
+    }
+
+    Shutdown.ConnHandle = ConnData->Connection.Handle;
+
+    uint32_t SecFlags = QUIC_CERTIFICATE_FLAG_DISABLE_CERT_VALIDATION;
+    Status =
+        MsQuic->SetParam(
+            ConnData->Connection,
+            QUIC_PARAM_LEVEL_CONNECTION,
+            QUIC_PARAM_CONN_CERT_VALIDATION_FLAGS,
+            sizeof(SecFlags),
+            &SecFlags);
+
+    if (QUIC_FAILED(Status)) {
+        WriteOutput("Failed Cert Validation Disable 0x%x\n", Status);
+        return Status;
+    }
+
+    if (!UseSendBuffer) {
+        BOOLEAN Opt = FALSE;
+        Status =
+            MsQuic->SetParam(
+                ConnData->Connection,
+                QUIC_PARAM_LEVEL_CONNECTION,
+                QUIC_PARAM_CONN_SEND_BUFFERING,
+                sizeof(Opt),
+                &Opt);
+
+        if (QUIC_FAILED(Status)) {
+            WriteOutput("Failed Disable Send Buffering 0x%x\n", Status);
+            return Status;
+        }
+    }
+
+    if (QuicAddrGetFamily(&LocalIpAddr) != AF_UNSPEC) {
+        MsQuic->SetParam(
+            ConnData->Connection,
+            QUIC_PARAM_LEVEL_CONNECTION,
+            QUIC_PARAM_CONN_LOCAL_ADDRESS,
+            sizeof(LocalIpAddr),
+            &LocalIpAddr);
+    }
+
+    Status =
+        MsQuic->ConnectionStart(
+            ConnData->Connection,
+            RemoteFamily,
+            TargetData.get(),
+            Port);
+
+    if (QUIC_FAILED(Status)) {
+        WriteOutput("Failed ConnectionStart 0x%x\n", Status);
+        return Status;
+    }
+
+    StreamData* StrmData = StreamDataAllocator.Alloc(this, ConnData->Connection);
+
+    Status =
+        MsQuic->StreamOpen(
+            ConnData->Connection,
+            QUIC_STREAM_OPEN_FLAG_UNIDIRECTIONAL,
+            [](HQUIC Handle, void* Context, QUIC_STREAM_EVENT* Event) -> QUIC_STATUS {
+                return ((StreamData*)Context)->Client->
+                    StreamCallback(
+                        Handle,
+                        Event,
+                        (StreamData*)Context);
+            },
+            StrmData,
+            &StrmData->Stream.Handle);
+
+    if (QUIC_FAILED(Status)) {
+        WriteOutput("Failed StreamOpen 0x%x\n", Status);
+        StreamDataAllocator.Free(StrmData);
+        return Status;
+    }
+
+    Shutdown.StreamHandle = StrmData->Stream.Handle;
```

Comment:
You shouldn't actually need to shutdown the stream explicitly; just the connection. The caveat to this: if the stream hasn't been successfully started, you need to just close it.